### PR TITLE
[TILES-COLLAB-2] PLU-238 edit collaborators endpoints

### PR DESCRIPTION
**Pull request by pregnantboy**
## Edit tile collaborator 
- Upsert
- Delete

  | Owner | Editor | Viewer | Shareable link
-- | -- | -- | -- | --
Can view tile | ✅ | ✅ | ✅ | ✅
Can view collaborators | ✅ | ✅ | ✅ |  ✅
Can view shareable link | ✅ | ✅ | ✅ |  ✅
Can edit data | ✅ | ✅ |   |  
Can rename pipe | ✅ | ✅ |   |  
Can adjust/edit columns | ✅ | ✅ |   |  
Can modify collaborators | ✅ | ✅ |   |  
Can regen shareable link | ✅ | ✅ |   |  
Can use in pipe | ✅ |   |   |  
Can delete table | ✅ |   |   |  

## Testing
- Unit tests added
- To manually test, use /graphql endpoint to manually insert, update and delete

### CODE REVIEW

File: packages/backend/src/graphql/mutations/tiles/upsert-table-collaborator.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    if (context.currentUser.email === validatedEmail) {
+      throw new BadUserInputError('Cannot change own role')
+    }
+
+    if (role === 'owner') {
+      throw new BadUserInputError('Cannot set collaborator role as owner')
+    }
+
+    await TableCollaborator.hasAccess(context.currentUser.id, tableId, 'editor')
```
Comment (m0nggh) [inline packages/backend/src/graphql/mutations/tiles/upsert-table-collaborator.ts:32]: nit: should this be checked first similar to `deleteTableCollaborator`?

File: packages/backend/src/graphql/mutations/tiles/delete-table-collaborator.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    if (validatedEmail === context.currentUser.email) {
+      throw new BadUserInputError('Cannot remove yourself')
+    }
+
+    const collaboratorUser = await User.query()
+      .findOne({
+        email: validatedEmail,
+      })
+      .throwIfNotFound()
```
Comment (m0nggh) [inline packages/backend/src/graphql/mutations/tiles/delete-table-collaborator.ts]:
```suggestion
      .throwIfNotFound('Collaborator email does not exist')
```

File: packages/backend/src/graphql/mutations/tiles/delete-table-collaborator.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    const validatedEmail = await validateAndParseEmail(email)
+    if (!validatedEmail) {
+      throw new BadUserInputError('Invalid collaborator email')
+    }
+
+    if (validatedEmail === context.currentUser.email) {
+      throw new BadUserInputError('Cannot remove yourself')
+    }
+
+    const collaboratorUser = await User.query()
```
Comment (m0nggh) [inline packages/backend/src/graphql/mutations/tiles/delete-table-collaborator.ts]: nit: can "early" block the mutation if the `collaboratorUser` is an owner here? (actually more of a suggestion)

Comment (m0nggh) [reply packages/backend/src/graphql/mutations/tiles/delete-table-collaborator.ts]: actually after poking, feels like we should not return true when the `collaboratorUser` is not a collaborator or if he is the owner?

**Review by m0nggh — CHANGES_REQUESTED** [review]
left some comments, and idk where to add this in the code but one more important thing is that right now, I can still view the tile if I'm deleted from the `table_collaborator`, need add a check in `get-tables` mutation: `.whereNull('table_collaborators.deleted_at')` (can ignore this if you made this change in another PR!)

and also in `table-collaborators` model, the `deletedAt` property in the `jsonSchema` should allow soft deletes now?

tested:
current owner can add new editor and re-add the same editor after deleting

need to test again:
current owner/editor should receive an error if deleting an owner or a non-collaborator?

File: packages/backend/src/graphql/mutations/tiles/upsert-table-collaborator.ts (hunk 1 above)
Comment (pregnantboy) [reply packages/backend/src/graphql/mutations/tiles/upsert-table-collaborator.ts:32]: checking bad inputs that dont need db queries first before checking permissions so the short circuiting will be faster

**Comment by pregnantboy** [comment]
* **#603** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/603?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#602** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/602?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#601** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/601?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#600** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/600?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#599** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/599?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#597** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/597?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `develop-v2`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @pregnantboy and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by m0nggh — APPROVED** [review]
lgtm sir

**Pull request merged**